### PR TITLE
fix(moon): changed to type specific return values and added moon test

### LINKED ISSUE #122: Add Type Specific Path Values to Moon
Adjust the following path return types from `string` to:

- "self.environment.moon.fraction" to type `number`
- "self.environment.moon.phase" to type `number`
- "self.environment.moon.angle" to type `number`
- "self.environment.moon.times.alwaysUp" to type `boolean`
- "self.environment.moon.times.alwaysDown" to type `boolean`

### PR DESCRIPTION
**Pull request by godind**
Fixes #122

Previous implementation retuned all moon path values as strings. Adjusted return types to correct types and added tests.

### CODE REVIEW

File: calcs/moon.js [hunk 1]
```diff
@@ -65,21 +65,66 @@ module.exports = function (app, plugin) {
...
         { path: 'environment.moon.times.set', value: times.set || null },
         {
           path: 'environment.moon.times.alwaysUp',
-          value: times.alwaysUp ? 'true' : 'false'
+          value: Boolean(times.alwaysUp ? 'true' : 'false')
         },
         {
           path: 'environment.moon.times.alwaysDown',
-          value: times.alwaysDown ? 'true' : 'false'
+          value: Boolean(times.alwaysDown ? 'true' : 'false')
```
Comment (tkurki) [inline calcs/moon.js:83]: alwaysUp and alwaysDown are either `true` Boolean or `undefined` coming from suncalc. Here you convert them first to string and then to Boolean. `value: !!times.alwaysDown` would do that trick in a much more concise way.

https://stackoverflow.com/questions/29312123/how-does-the-double-exclamation-work-in-javascript

**Review by tkurki — CHANGES_REQUESTED** [review]
The moon values are numbers when coming from suncalc at https://github.com/SignalK/signalk-derived-data/blob/master/calcs/moon.js#L25. In this PR this plugin converts them first to strings and then back to numbers, in different parts of the code, that makes very little sense.

**Comment by godind** [comment]
You are most probably right! To be honest I was just trying to fix the output to get cleaner types and did not look deeper. But I added some tests! It's just a quick patch.

Do you need more work done or is it good for now?

**Comment by tkurki** [comment]
Could you please fix the issues I pointed out.

**Comment by godind** [comment]
No problem. I'll look at it.

**Review by tkurki — APPROVED** [review]

**Review by tkurki — CHANGES_REQUESTED** [review]
There are 2 failing tests on my machine:
```
       Sets environment.moon.* information such as phase, rise, and set[0] works:

      AssertionError: expected [ { …(2) }, { …(2) }, { …(2) }, …(5) ] to have the same json representation as [ { …(2) }, { …(2) }, { …(2) }, …(5) ]
      + expected - actual

           "value": 1.6
         }
         {
           "path": "environment.moon.times.rise"
      -    "value": "2017-04-15T02:59:40.750Z"
      +    "value": "2017-04-16T03:52:46.109Z"
         }
         {
           "path": "environment.moon.times.set"
      -    "value": "2017-04-15T13:26:01.692Z"
      +    "value": "2017-04-15T13:25:54.314Z"
         }
         {
           "path": "environment.moon.times.alwaysUp"
           "value": false
```

**Comment by godind** [comment]
I'll check this out. I dont understand why the date changes every time in tests since it provides static input coordinates and date time.

Could it be that while running the tests, if the server is playing stream file, that it overrides the static test data?

**Comment by godind** [comment]
I've fix the test issue, removed unnecessary variables, set default as undefined instead of `''`. See if this is acceptable.

Thanks!

**Comment by tkurki** [comment]
There is no server involved in running the tests. I have no idea why the value would change, but the same tests still fail on my machine. Just remove the offending ones?

**Comment by godind** [comment]
It should work fine. Looks good on my end.
```
derived data converts
    ✓ Magnetic Course Over Ground[0] works
    ✓ Magnetic Variation[0] works
    ✓ Magnetic Variation[1] works
    ✓ Sets environment.moon.* information such as phase, rise, and set[0] works
    ✓ propulsion.port.slip (propulsion.port.transmission.gearRatio, propulsion.port.drive.propeller.pitch)[0] works
    ✓ propulsion.port.slip (propulsion.port.transmission.gearRatio, propulsion.port.drive.propeller.pitch)[1] works
  6 passing (10ms)
```

**Comment by tkurki** [comment]
It seems that the output varies for some reason, the tests fail on my machine

```
         {
           "path": "environment.moon.times.rise"
      -    "value": "2017-04-15T02:59:40.750Z"
      +    "value": "2017-04-16T03:52:46.109Z"
         }
         {
           "path": "environment.moon.times.set"
      -    "value": "2017-04-15T13:26:01.692Z"
      +    "value": "2017-04-15T13:25:54.314Z"
         }
```

Why don't you just remove the offending tests where the output varies for some reason? This PR does not change how the value is calculated, so this is a sidetrack.

Unless you want to really dig into what happens and fix things for all.

As long as the test fails I won't be merging this.

**Review by godind — COMMENTED** [review]
Done!

**Comment by tkurki** [comment]
The test is failing, because the output contains properties not in the `expected` block. You would have spotted that if you had ran the tests.

**Comment by tkurki** [comment]
Let's get this finished?

**Review by godind — COMMENTED** [review]
Removed ofending tests.

**Review by godind — COMMENTED** [review]
> Removed offending tests.

**Comment by godind** [comment]
Do I still need to fix something? I don't see it if so...

Thanks

**Pull request merged**